In [1]:
# import 'Pandas'  # data processing, CSV file I/O (e.g. pd.read_csv)
import pandas as pd 

# import 'Numpy' # linear algebra
import numpy as np

# import subpackage of Matplotlib
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap

# import 'Seaborn' 
import seaborn as sns

# to suppress warnings 
from warnings import filterwarnings
filterwarnings('ignore')

# display all columns of the dataframe
pd.options.display.max_columns = None

# display all rows of the dataframe
pd.options.display.max_rows = None
 
# to display the float values upto 6 decimal places     
pd.options.display.float_format = '{:.6f}'.format

# import train-test split 
from sklearn.model_selection import train_test_split

# to visualize missing values 
import missingno as msn


# import various functions from statsmodels
import statsmodels
import statsmodels.api as sm
from scipy.stats import shapiro , pearsonr

# import StandardScaler to perform scaling
from sklearn.preprocessing import StandardScaler 

# import various functions from sklearn 
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score
from sklearn import metrics
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier

from sklearn.tree import DecisionTreeClassifier,plot_tree
from sklearn.metrics import classification_report
from sklearn.metrics import cohen_kappa_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import roc_curve
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score

# import function to perform feature selection
from sklearn.feature_selection import RFE
from xgboost import XGBClassifier

In [2]:
def load_data(x):
    df = pd.read_csv('x')
    return df



def miss_val(x):
    miss_val =[]
    for i in x.columns:
        for j in x[i]:
            if x[x[i].isna()].shape[0]>=1:
                miss_val.extend(x[x[i].isna()].index)
    return miss_val

            
            
def univariate_outlier(x):
    q1 = df[x].describe().T["25%"]
    q3 = df[x].describe().T["75%"]
    iqr = q3-q1
    inner = q1 -1.5*iqr
    outer = q3 +1.5*iqr
    out=[]
    for i,j in df[x].iteritems():
        if inner<j<outer:
            pass
        else:
            out.append(j)
    return out

    
    
def feature_box_kde_normality(x):
    normality_test(df[x])
    out=univariate_outlier(x)
    if len(out)==0:
        print(f"There is no outlier in {x}")
    else:
        print("Outlier :",out)
    f,(ax1,ax2) = plt.subplots(1,2,figsize=[15,3])
    sns.boxplot(data=df,x=x,ax=ax1)
    sns.kdeplot(data=df,x=x,ax=ax2)
    
def normality_test(x):
    s,p = stats.shapiro(x)
    print(f"The statistic value is {s} with p-value : {p}")
    if p>0.05:
        print(f"{x.name} is normally distributed")
    else:
        print(f"{x.name} is not normally distributed")
        
def missing_values(df):
    for i in df.columns:
        df[i].fillna(df[i].mode()[0],inplace=True)
        
def feature_analysis(x):
    a = df[x].unique()
    d = df[x].value_counts()
    b = df[x].value_counts(normalize = True)*100
    c = df[x].isna().sum()
    print('unique_values:', '\n', a, '\n', 'value_counts:','\n', d, '\n','proportion of value:', '\n', b, '\n','count of nan values:', c)
    
def cat(x):
    sns.countplot(data= df, x = col)
    plt.show()
    print(df1[x]. value_counts().head(1))

def high_frequency(x):
        # set the figure size
        plt.rcParams["figure.figsize"] = [10,8]
        # plot the bar plot
        ax= sns.countplot(x= df[x], data = df)
        # rotate the x-axis labels
        ax.set_xticklabels(ax.get_xticklabels(),rotation=30)
# display the values on the bar plot
# annotate(): it is use to emphasize specific observations
# get_height(): return the height of the bar 
# (p.get_x() + p.get_width() / 2., p.get_height()): it is use to place the text on the top
# get_width() and get_x(): return the value of the x-axis
# ha = 'center': used for horizontal alignment
#  xytext(): the text appear at the location given by xytext
# textcoords: specify an offset (in points) from the xy value
        for p in ax.patches:           # use for loop to annotate each bin of the plot
            ax.annotate(format(p.get_height(), '.2f'), (p.get_x() + p.get_width() / 2., p.get_height()), ha = 'center',
                   xytext = (0, 15), textcoords = 'offset points')
        print('the variable with more frequency is', df[x].mode()[0])

    
def comparison_plot(x,y):
    plt.rcParams["figure.figsize"] = [15,10]
    ax = sns.countplot(x=df[x], hue= y,data=df)
    for p in ax.patches:
        ax.annotate(format(p.get_height(), '.2f'), (p.get_x() + p.get_width() / 2., p.get_height()), ha = 'center',
                   xytext = (0, 15), textcoords = 'offset points')
    plt.show()
    
def multi_comparison (x,y,z):
    fig, (axis1)  = plt.subplots(1, 1,figsize=(15,10))
    sns.barplot(x=df[x], y= df[y], hue=df[z], data=df, ax=axis1)
    

def cat_dtype_change(x):
    df[x]= df[x].astype('object')
    df[x].info()
    

def count_pie(x):
    plt.figure(figsize = (15,5))
    plt.subplot(121)
    df[x].value_counts().plot(kind ='pie',autopct = '%.2f')
    plt.subplot(122)
    sns.countplot(data = df, x = df[x])
    print('the variable with more frequency is', df[x].mode()[0])
    
def multi_pie_count(x,z,y):
    plt.figure(figsize = (20,5))
    plt.subplot(131)
    df[x].value_counts().plot(kind ='pie',autopct = '%.2f')
    plt.subplot(132)
    sns.countplot(data = df, x = df[x],hue =df[z])
    plt.subplot(133)
    sns.countplot(data = df, x = df[x],hue =df[y])
    
    
def train_test(x,y):
    x_encoded = pd.get_dummies(x, drop_first=True)
    scaler = StandardScaler()
    scaler.fit(x_encoded)
    x1 = scaler.fit_transform(x_encoded)
    x_train, x_test, y_train, y_test = train_test_split(x1, y, test_size=0.3, random_state=45)
    
def results( x_test, y_test,x_train, y_train):
    Logit_Model = LogisticRegression()
    Logit_Model.fit(x_train,y_train)
    decision_tree = DecisionTreeClassifier(criterion = 'entropy')
    decision_tree.fit(x_train,y_train)
    random_forest = RandomForestClassifier(n_estimators=10)
    random_forest.fit(x_train, y_train)
    from sklearn.ensemble import AdaBoostClassifier
    ad= AdaBoostClassifier(n_estimators = 40, random_state = 10) #creation of the object
    ad.fit(x_train, y_train)
    from sklearn.ensemble import GradientBoostingClassifier
    gboost_model = GradientBoostingClassifier(n_estimators = 100, max_depth = 3, random_state = 8)
    gboost_model.fit(x_train, y_train)
    from xgboost import XGBClassifier
    xgb_clf = XGBClassifier()
    xgb_clf.fit(x_train, y_train)
    from sklearn.naive_bayes import GaussianNB
    gnb = GaussianNB()
    gnb.fit(x_train, y_train)
    models = {'logis': Logit_Model, 'decision tree': decision_tree,'random_forest': random_forest, 
               'Adaboost': ad, 'XG Boost': xgb_clf, 'Gradient Boosting' : gboost_model, 'Naive Bayes': gnb }
    for model_name , model in models.items():
        y_pred = model.predict(x_test)
          # Compute the evaluation metrics
        accuracy = accuracy_score(y_test, y_pred)
        f1 = f1_score(y_test, y_pred,average='micro')
        precision = precision_score(y_test, y_pred,average='micro')
        recall = recall_score(y_test, y_pred,average='micro')
    
        result = pd.DataFrame({'Model Name' : model_name,
                            'Accuracy': [accuracy],
                            'F1 Score': [f1],
                            'Precision': [precision],
                            'Recall': [recall]})
    return result


def confusion_Matrix(x_train, x_test, y_train, y_test, model_name):
        train_pred = xgb_model.predict(x_train)
        print('training results', model_name)
        print(classification_report(y_train, train_pred))
        cm= confusion_matrix(y_train, train_pred)
        print('Confusion matrix')
        print('/n')
        print(cm)
    
        print("---------------------------------------------------")
        print('Test Results')
        test_pred = xgb_model.predict(x_test)
        print(classification_report(y_test, test_pred))
        cm =confusion_matrix(y_test, test_pred)
        print('Confusion matrix')
        print(cm)
    

# Importing required datasets

In [3]:
# importing datasets with imputed values for model building

import pandas as pd
import easygui
path= easygui.fileopenbox()
df_imputer=pd.read_csv(path)
df_imputer.head()

Gender      Age  Occupation  City_Category  Stay_In_Current_City_Years  \
0 0.000000 0.000000   10.000000       0.000000                    2.000000   
1 0.000000 0.000000   10.000000       0.000000                    2.000000   
2 0.000000 0.000000   10.000000       0.000000                    2.000000   
3 0.000000 0.000000   10.000000       0.000000                    2.000000   
4 1.000000 6.000000   16.000000       2.000000                    4.000000   

   Marital_Status  Product_Category_1  Product_Category_2  Product_Category_3  \
0        0.000000            3.000000            4.000000            5.000000   
1        0.000000            1.000000            6.000000           14.000000   
2        0.000000           12.000000           16.400000           16.500000   
3        0.000000           12.000000           14.000000           16.000000   
4        0.000000            8.000000           14.000000           17.000000   

      Purchase  product fre  clus_labels Purchase_ranges  
0  8370.000000     0.000000            2         Range 3  
1 15200.000000     0.000000            4         Range 4  
2  1422.000000     0.000000            5         Range 1  
3  1057.000000     0.000000            3         Range 1  
4  7969.000000     0.000000            1         Range 2

In [6]:
# importing clustering dataset which we done 
import pandas as pd
import easygui
path= easygui.fileopenbox()
clustering=pd.read_csv(path)
clustering.head()

Gender      Age  Occupation  City_Category  Stay_In_Current_City_Years  \
0 0.000000 0.000000   10.000000       0.000000                    2.000000   
1 0.000000 0.000000   10.000000       0.000000                    2.000000   
2 0.000000 0.000000   10.000000       0.000000                    2.000000   
3 0.000000 0.000000   10.000000       0.000000                    2.000000   
4 1.000000 6.000000   16.000000       2.000000                    4.000000   

   Marital_Status  Product_Category_1  Product_Category_2  Product_Category_3  \
0        0.000000            3.000000            4.000000            5.000000   
1        0.000000            1.000000            6.000000           14.000000   
2        0.000000           12.000000           16.400000           16.500000   
3        0.000000           12.000000           14.000000           16.000000   
4        0.000000            8.000000           14.000000           17.000000   

   Purchase  product fre Purchase_cat  clus label  
0      8370     0.000000      Range 3           2  
1     15200     0.000000      Range 4           4  
2      1422     0.000000      Range 1           5  
3      1057     0.000000      Range 1           3  
4      7969     0.000000      Range 2           1

In [7]:
# importing datasets with imputed values for model building

import pandas as pd
import easygui
path= easygui.fileopenbox()
df_mice=pd.read_csv(path)
df_mice.head()

Gender      Age  Occupation  City_Category  Stay_In_Current_City_Years  \
0 0.000000 0.000000   10.000000       0.000000                    2.000000   
1 0.000000 0.000000   10.000000       0.000000                    2.000000   
2 0.000000 0.000000   10.000000       0.000000                    2.000000   
3 0.000000 0.000000   10.000000       0.000000                    2.000000   
4 1.000000 6.000000   16.000000       2.000000                    4.000000   

   Marital_Status  Product_Category_1  Product_Category_2  Product_Category_3  \
0        0.000000            3.000000            4.000000           12.000000   
1        0.000000            1.000000            6.000000           14.000000   
2        0.000000           12.000000           15.000000           16.000000   
3        0.000000           12.000000           14.000000           16.000000   
4        0.000000            8.000000           16.000000           17.000000   

      Purchase  product fre  
0  8370.000000     0.000413  
1 15200.000000     0.001056  
2  1422.000000     0.000185  
3  1057.000000     0.000620  
4  7969.000000     0.000369

In [8]:
# importing datasets with imputed values for model building

import pandas as pd
import easygui
path= easygui.fileopenbox()
df_rf=pd.read_csv(path)
df_rf.head()

Gender      Age  Occupation  City_Category  Stay_In_Current_City_Years  \
0 0.000000 0.000000   10.000000       0.000000                    2.000000   
1 0.000000 0.000000   10.000000       0.000000                    2.000000   
2 0.000000 0.000000   10.000000       0.000000                    2.000000   
3 0.000000 0.000000   10.000000       0.000000                    2.000000   
4 1.000000 6.000000   16.000000       2.000000                    4.000000   

   Marital_Status  Product_Category_1  Product_Category_2  Product_Category_3  \
0        0.000000            3.000000            4.000000            5.000000   
1        0.000000            1.000000            6.000000           14.000000   
2        0.000000           12.000000           16.400000           16.500000   
3        0.000000           12.000000           14.000000           16.000000   
4        0.000000            8.000000           14.000000           17.000000   

      Purchase  product fre  
0  8370.000000     0.000000  
1 15200.000000     0.000000  
2  1422.000000     0.000000  
3  1057.000000     0.000000  
4  7969.000000     0.000000

# 1. BUILD A MODEL WITH ``df_imputer`` DATASET

In [9]:
x= df_imputer

y =clustering['clus label']

In [10]:
x_train, x_test, y_train, y_test = train_test_split(x, y, random_state = 142, test_size = 0.3)
print('Shape of the X_train', x_train.shape)
print('Shape of the y_train', y_train.shape)

print('Shape of the X_test', x_test.shape)
print('Shape of the y_test', y_test.shape)

Shape of the X_train (385047, 11)
Shape of the y_train (385047,)
Shape of the X_test (165021, 11)
Shape of the y_test (165021,)


In [29]:
def results( x_test, y_test,x_train, y_train):
    Logit_Model = LogisticRegression()
    Logit_Model.fit(x_train,y_train)
    decision_tree = DecisionTreeClassifier(criterion = 'entropy')
    decision_tree.fit(x_train,y_train)
    random_forest = RandomForestClassifier(n_estimators=10)
    random_forest.fit(x_train, y_train)
    from sklearn.ensemble import AdaBoostClassifier
    ad= AdaBoostClassifier(n_estimators = 40, random_state = 10) #creation of the object
    ad.fit(x_train, y_train)
    from sklearn.ensemble import GradientBoostingClassifier
    gboost_model = GradientBoostingClassifier(n_estimators = 100, max_depth = 3, random_state = 8)
    gboost_model.fit(x_train, y_train)
    from xgboost import XGBClassifier
    xgb_clf = XGBClassifier()
    xgb_clf.fit(x_train, y_train)
    from sklearn.naive_bayes import GaussianNB
    gnb = GaussianNB()
    gnb.fit(x_train, y_train)
    models = {'logis': Logit_Model, 'decision tree': decision_tree,'random_forest': random_forest, 
               'Adaboost': ad, 'XG Boost': xgb_clf, 'Gradient Boosting' : gboost_model, 'Naive Bayes': gnb }
    result = pd.DataFrame(columns=['Model', 'Accuracy', 'F1 Score', 'Precision', 'Recall'])
    for model_name , model in models.items():
        y_pred = model.predict(x_test)
        accuracy = accuracy_score(y_test, y_pred)
        f1 = f1_score(y_test, y_pred,average='micro')
        precision = precision_score(y_test, y_pred,average='micro')
        recall = recall_score(y_test, y_pred,average='micro')
        result.loc[model_name] = [type(model).__name__, accuracy, f1, precision, recall]

    return result

In [30]:
results( x_test, y_test,x_train, y_train)

Model  Accuracy  F1 Score  Precision  \
logis                      LogisticRegression  0.362584  0.362584   0.362584   
decision tree          DecisionTreeClassifier  0.870835  0.870835   0.870835   
random_forest          RandomForestClassifier  0.868417  0.868417   0.868417   
Adaboost                   AdaBoostClassifier  0.727180  0.727180   0.727180   
XG Boost                        XGBClassifier  0.924113  0.924113   0.924113   
Gradient Boosting  GradientBoostingClassifier  0.830931  0.830931   0.830931   
Naive Bayes                        GaussianNB  0.519607  0.519607   0.519607   

                    Recall  
logis             0.362584  
decision tree     0.870835  
random_forest     0.868417  
Adaboost          0.727180  
XG Boost          0.924113  
Gradient Boosting 0.830931  
Naive Bayes       0.519607

# ``Interpretation`` 

#  Our target column is highly imbalanced. So we are considered ``F1-SCORE``.

``METRICS``


``ACCURACY`` Accuracy measures the proportion of correct predictions over the total number of predictions, and it is a good metric when the classes are balanced (i.e., each profit range has roughly the same number of instances).

``PRECISION``   Precision measures the proportion of true positive predictions over the total number of positive predictions, and it is a good metric when the cost of a false positive prediction is high.

``RECALL``    Recall measures the proportion of true positive predictions over the total number of actual positive instances, and it is a good metric when the cost of a false negative prediction is high. 

``F1-SCORE``  F1 score is the harmonic mean of precision and recall, and it is a good metric when you want to balance the trade-off between precision and recall.

             

# 2. BUILD A MODEL WITH ``df_mice`` DATASET

In [31]:
x1= df_mice

y1 =clustering['clus label']

In [32]:
x_train, x_test, y_train, y_test = train_test_split(x1, y1, random_state = 142, test_size = 0.3)
print('Shape of the X_train', x_train.shape)
print('Shape of the y_train', y_train.shape)

print('Shape of the X_test', x_test.shape)
print('Shape of the y_test', y_test.shape)

Shape of the X_train (385047, 11)
Shape of the y_train (385047,)
Shape of the X_test (165021, 11)
Shape of the y_test (165021,)


In [33]:
results( x_test, y_test,x_train, y_train)

Model  Accuracy  F1 Score  Precision  \
logis                      LogisticRegression  0.362584  0.362584   0.362584   
decision tree          DecisionTreeClassifier  0.870798  0.870798   0.870798   
random_forest          RandomForestClassifier  0.871368  0.871368   0.871368   
Adaboost                   AdaBoostClassifier  0.727180  0.727180   0.727180   
XG Boost                        XGBClassifier  0.924113  0.924113   0.924113   
Gradient Boosting  GradientBoostingClassifier  0.830931  0.830931   0.830931   
Naive Bayes                        GaussianNB  0.519607  0.519607   0.519607   

                    Recall  
logis             0.362584  
decision tree     0.870798  
random_forest     0.871368  
Adaboost          0.727180  
XG Boost          0.924113  
Gradient Boosting 0.830931  
Naive Bayes       0.519607

# 3.BUILD A MODEL WITH ``df_rf`` DATASET

In [34]:
x2= df_rf

y2 =clustering['clus label']

In [35]:
x_train, x_test, y_train, y_test = train_test_split(x2, y2, random_state = 142, test_size = 0.3)
print('Shape of the X_train', x_train.shape)
print('Shape of the y_train', y_train.shape)

print('Shape of the X_test', x_test.shape)
print('Shape of the y_test', y_test.shape)

Shape of the X_train (385047, 11)
Shape of the y_train (385047,)
Shape of the X_test (165021, 11)
Shape of the y_test (165021,)


In [36]:
results( x_test, y_test,x_train, y_train)

Model  Accuracy  F1 Score  Precision  \
logis                      LogisticRegression  0.368723  0.368723   0.368723   
decision tree          DecisionTreeClassifier  0.961987  0.961987   0.961987   
random_forest          RandomForestClassifier  0.962071  0.962071   0.962071   
Adaboost                   AdaBoostClassifier  0.789021  0.789021   0.789021   
XG Boost                        XGBClassifier  0.997982  0.997982   0.997982   
Gradient Boosting  GradientBoostingClassifier  0.941129  0.941129   0.941129   
Naive Bayes                        GaussianNB  0.533302  0.533302   0.533302   

                    Recall  
logis             0.368723  
decision tree     0.961987  
random_forest     0.962071  
Adaboost          0.789021  
XG Boost          0.997982  
Gradient Boosting 0.941129  
Naive Bayes       0.533302

``Interpretation``

         1. From the Imputer dataset Xtreme Gradient Boosting is best with F1-Score ``0.924113``
    
         2. From the Mice dataset Xtreme Gradient Boosting is best with F1-Score ``0.924113``
    
         3. From the Random Forest dataset Xtreme Gradient Boosting is best with F1-Score ``0.997982``

# BASE MODEL  - ``Xtreme Gradient Boosting Classifier``

From the above model Random Forest imputer method works best so consider that dataset as a Final Dataset and a                 ``Xtreme Gradient Boosting Classifier`` considered as a ``Base Model``

In [38]:
clustering.columns

Index(['Gender', 'Age', 'Occupation', 'City_Category',
       'Stay_In_Current_City_Years', 'Marital_Status', 'Product_Category_1',
       'Product_Category_2', 'Product_Category_3', 'Purchase', 'product fre',
       'Purchase_cat', 'clus label'],
      dtype='object')

In [40]:
df_rf['clus_labels'] = clustering['clus label']
df_rf['Purchase_ranges']= clustering['Purchase_cat']
df_rf.head()

Gender      Age  Occupation  City_Category  Stay_In_Current_City_Years  \
0 0.000000 0.000000   10.000000       0.000000                    2.000000   
1 0.000000 0.000000   10.000000       0.000000                    2.000000   
2 0.000000 0.000000   10.000000       0.000000                    2.000000   
3 0.000000 0.000000   10.000000       0.000000                    2.000000   
4 1.000000 6.000000   16.000000       2.000000                    4.000000   

   Marital_Status  Product_Category_1  Product_Category_2  Product_Category_3  \
0        0.000000            3.000000            4.000000            5.000000   
1        0.000000            1.000000            6.000000           14.000000   
2        0.000000           12.000000           16.400000           16.500000   
3        0.000000           12.000000           14.000000           16.000000   
4        0.000000            8.000000           14.000000           17.000000   

      Purchase  product fre  clus_labels Purchase_ranges  
0  8370.000000     0.000000            2         Range 3  
1 15200.000000     0.000000            4         Range 4  
2  1422.000000     0.000000            5         Range 1  
3  1057.000000     0.000000            3         Range 1  
4  7969.000000     0.000000            1         Range 2

In [41]:
# Take the copy of the final dataset 
df_rf.to_csv('C:\\Users\\BIS\\Documents\\df_sales_final.csv', index = False)

# Train and Test split 

In [5]:
import pandas as pd
import easygui
path= easygui.fileopenbox()
df_rf=pd.read_csv(path)
df_rf.head()

Gender      Age  Occupation  City_Category  Stay_In_Current_City_Years  \
0 0.000000 0.000000   10.000000       0.000000                    2.000000   
1 0.000000 0.000000   10.000000       0.000000                    2.000000   
2 0.000000 0.000000   10.000000       0.000000                    2.000000   
3 0.000000 0.000000   10.000000       0.000000                    2.000000   
4 1.000000 6.000000   16.000000       2.000000                    4.000000   

   Marital_Status  Product_Category_1  Product_Category_2  Product_Category_3  \
0        0.000000            3.000000            4.000000            5.000000   
1        0.000000            1.000000            6.000000           14.000000   
2        0.000000           12.000000           16.400000           16.500000   
3        0.000000           12.000000           14.000000           16.000000   
4        0.000000            8.000000           14.000000           17.000000   

      Purchase  product fre  clus_labels Purchase_ranges  
0  8370.000000     0.000000            2         Range 3  
1 15200.000000     0.000000            4         Range 4  
2  1422.000000     0.000000            5         Range 1  
3  1057.000000     0.000000            3         Range 1  
4  7969.000000     0.000000            1         Range 2

In [6]:
X= df_rf.drop(['Purchase_ranges', 'clus_labels'],axis = 1)

Y =df_rf['clus_labels']

In [7]:
X.head()

Gender      Age  Occupation  City_Category  Stay_In_Current_City_Years  \
0 0.000000 0.000000   10.000000       0.000000                    2.000000   
1 0.000000 0.000000   10.000000       0.000000                    2.000000   
2 0.000000 0.000000   10.000000       0.000000                    2.000000   
3 0.000000 0.000000   10.000000       0.000000                    2.000000   
4 1.000000 6.000000   16.000000       2.000000                    4.000000   

   Marital_Status  Product_Category_1  Product_Category_2  Product_Category_3  \
0        0.000000            3.000000            4.000000            5.000000   
1        0.000000            1.000000            6.000000           14.000000   
2        0.000000           12.000000           16.400000           16.500000   
3        0.000000           12.000000           14.000000           16.000000   
4        0.000000            8.000000           14.000000           17.000000   

      Purchase  product fre  
0  8370.000000     0.000000  
1 15200.000000     0.000000  
2  1422.000000     0.000000  
3  1057.000000     0.000000  
4  7969.000000     0.000000

In [8]:
Y.head()

0    2
1    4
2    5
3    3
4    1
Name: clus_labels, dtype: int64

In [9]:
x_train, x_test, y_train, y_test = train_test_split(X, Y, random_state = 142, test_size = 0.3)
print('Shape of the X_train', x_train.shape)
print('Shape of the y_train', y_train.shape)

print('Shape of the X_test', x_test.shape)
print('Shape of the y_test', y_test.shape)

Shape of the X_train (385047, 11)
Shape of the y_train (385047,)
Shape of the X_test (165021, 11)
Shape of the y_test (165021,)


# Model

In [52]:
xgb_model = XGBClassifier(learning_rate = 0.01, gamma = 2)

xgb_model.fit(x_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=2, gpu_id=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.01, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              n_estimators=100, n_jobs=None, num_parallel_tree=None,
              objective='multi:softprob', predictor=None, ...)

In [59]:
confusion_Matrix(x_train, x_test, y_train, y_test,'XGBoost_Classifier' )

training results XGBoost_Classifier
              precision    recall  f1-score   support

           0       0.80      0.82      0.81     95693
           1       0.82      0.84      0.83     89629
           2       0.81      0.70      0.75     45708
           3       0.77      0.78      0.77     54746
           4       0.80      0.80      0.80     52992
           5       0.80      0.78      0.79     46279

    accuracy                           0.80    385047
   macro avg       0.80      0.79      0.79    385047
weighted avg       0.80      0.80      0.80    385047

Confusion matrix
/n
[[78551  6994  1835  3981  1930  2402]
 [ 5715 75458  1744  2887  2516  1309]
 [ 2725  3679 32049  2270  2789  2196]
 [ 4986  1570  1246 42904  2100  1940]
 [ 3178  2683  1647  1769 42652  1063]
 [ 3559  2168  1181  2187  1109 36075]]
---------------------------------------------------
Test Results
              precision    recall  f1-score   support

           0       0.80      0.82      0.81   

#  HYPER PARAMETER TUNING

In [25]:
# Take a ample for tuning

df_sample = df_rf.sample(n=100000, random_state= 142)
X= df_sample.drop(['Purchase_ranges', 'clus_labels'],axis = 1)

Y =df_sample['clus_labels']

In [19]:
tuned_paramaters = [{
                     'learning_rate': [0.1, 0.01, 0.001], 
                     'gamma': [2, 3, 4],
                     'max_depth': [2, 3, 4]}]
 
xgb_model = XGBClassifier(random_state = 8, n_jobs=-1)

xgb_grid = GridSearchCV(estimator = xgb_model, 
                         param_grid = tuned_paramaters, 
                         cv = 5, 
                         n_jobs=-1)
xgb_grid.fit(X, Y)

print('Best parameters for Extreme Gradient Boositng Classifier: ', xgb_grid.best_params_, '\n')

Best parameters for Extreme Gradient Boositng Classifier:  {'gamma': 2, 'learning_rate': 0.1, 'max_depth': 4} 



In [13]:
# Build a model wirth best parameters 
xgb_model1 = XGBClassifier(learning_rate = 0.01, gamma = 2, max_depth = 4)

xgb_model1.fit(x_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=2, gpu_id=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.01, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=4, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              n_estimators=100, n_jobs=None, num_parallel_tree=None,
              objective='multi:softprob', predictor=None, ...)

In [14]:
y_pred_xgb = xgb_model1.predict(x_test)

In [24]:
accuracy = accuracy_score(y_test, y_pred_xgb)
f1 = f1_score(y_test, y_pred_xgb,average='micro')
precision = precision_score(y_test, y_pred_xgb,average='micro')
recall = recall_score(y_test, y_pred_xgb,average='micro')
print('Accuracy', accuracy, '\n', 'F1-Score', f1, '\n', 'precision', precision, '\n', 'Recall', recall)

Accuracy 0.7260045691154459 
 F1-Score 0.7260045691154458 
 precision 0.7260045691154459 
 Recall 0.7260045691154459


In [15]:
def confusion_Matrix(x_train, x_test, y_train, y_test, model_name):
        train_pred = xgb_model1.predict(x_train)
        print('training results', model_name)
        print(classification_report(y_train, train_pred))
        cm= confusion_matrix(y_train, train_pred)
        print('Confusion matrix')
        print('/n')
        print(cm)
    
        print("---------------------------------------------------")
        print('Test Results')
        test_pred = xgb_model1.predict(x_test)
        print(classification_report(y_test, test_pred))
        cm =confusion_matrix(y_test, test_pred)
        print('Confusion matrix')
        print(cm)
    

In [16]:
confusion_Matrix(x_train, x_test, y_train, y_test,'XGBoost_Classifier')

training results XGBoost_Classifier
              precision    recall  f1-score   support

           0       0.72      0.74      0.73     95693
           1       0.72      0.81      0.77     89629
           2       0.73      0.63      0.68     45708
           3       0.73      0.63      0.68     54746
           4       0.74      0.76      0.75     52992
           5       0.72      0.69      0.71     46279

    accuracy                           0.72    385047
   macro avg       0.73      0.71      0.72    385047
weighted avg       0.73      0.72      0.72    385047

Confusion matrix
/n
[[70674 12605  2241  3509  2871  3793]
 [ 7435 72992  1375  2254  3628  1945]
 [ 3258  5926 28847  1640  3478  2559]
 [ 9225  2528  3053 34406  2973  2561]
 [ 3554  3056  2351  2231 40213  1587]
 [ 4281  3857  1687  2977  1489 31988]]
---------------------------------------------------
Test Results
              precision    recall  f1-score   support

           0       0.72      0.73      0.73   